### DISCUSSION TITLE
由mail作业惨案引发的对于make的理解

### DISCUSSION THREAD

**Discussion by yuanyi-thu** [physics-data/faq#128] [General] [answered]
# 什么是make:

`make`命令的逻辑:
        
        1-进入Makefile
        2-看到第一个target
        3-递归定位到他的依赖链的每一个末端块
        4-对每一个块，执行以下逻辑

对于每一个块如下：
    
    A:B
        blabla

执行逻辑为：

    if A,B 都存在:
        if A 在 B 之后生成:
            那没事了 -> 结束
        else：#B在A之后生成
            执行blabla -> 结束
    elif A无 B有:
        执行blabla -> 结束
    elif B非空,但实际上找不到对应文件:
        if 本块为最底层:
            报错
        else:
            (此前一定已经执行过B的子块)
            执行blabla -> 结束
    else: # B 为空
        if A 无:
            执行blabla -> 结束
        else: # A 存在
            那没事了 -> 结束

一个块结束之后:
    
    if 本块是最高层:
        结束make
    else:
        进入上一级块
        对上一级块执行同上逻辑

（块正式一点的名字：
    
    target:source
        command

## 特别注意递归

mails作业中一种可能的常见情况是，对于中间块，A和B都非空且都无，根据上述流程，从子块上来之后，仍然判断需要执行一遍`blabla`，这将造成下述 `make -q` 过程中判定false

# 什么是make -q

执行`make`，若在任何一个块内判定需要进入`blabla`部分则返回1(False),否则返回0(True)

## 特例

make -q 返回正确的判据其实并不是严格的任何块都不能进入，如果某一个块的`blabla`部分是**纯空的**，那么即使判定进入命令体也不妨碍返回值。

**Comment by yuanyi-thu** [comment, marked as answer]
然而因为grader的clean不知道为什么不用自己的make clean，最后只能对答案采取了过拟合

**Reply by heroxbd** [reply]
https://github.com/physics-data/faq/discussions/124 提出了同样的问题。

**Comment by YiHui-Liu** [comment]
另：https://github.com/physics-data/faq/discussions/130#discussioncomment-1281739

**Reply by Harry-Chen** [reply]
是的，因为Makefile 还存在 phony target，不一定会经历对于依赖的检查
